Cu 86b2qw06r feat add upload event listeners

amjedidiah opened the pull request:
Added the following event listeners:

- onClick
- onUpload
- onUploadFail
- onCompletedUpload
- onAllCompleted
- onFileRemove
- onDragOver
- onDragLeave
- onDrop
- onFileProgress
- onTotalProgress

amjedidiah committed:
feat(uploader): enhanced file preview with react-file-viewer

- Added support for previewing multiple file types (PDF, DOCX, XLSX, etc.)
- Implemented `react-file-viewer` for better file preview capabilities
- Maintained fallback to FileIcon component when preview fails

OTHER CHANGES
- Resolved React version mismatch (16.14.0 -> 18.2.0) to fix hook errors in Storybook
- Added VSCode workspace folder to `.gitignore`

BREAKING CHANGE: New dependency `react-file-viewer` is required

amjedidiah committed:
feat(uploader): enhanced file preview hover interactions

- Added hover effect with background color change and shadow elevation
- Implemented cursor pointer for clickable file previews
- Added active state for click feedback
- Enhanced dark mode hover states
- Maintained consistent hover behaviour across mini and full preview modes

amjedidiah committed:
fix(uploader): resolved DOCX preview conflicts and improve type safety

- Added modal-based DOCX preview to prevent interference between multiple previews
- Created FileWithId interface and shared type definitions
- Fixed file ID generation and management
- Added proper cleanup for object URLs
- Ensured consistent key generation for previews

BREAKING CHANGE: File state type is now FileWithId[] instead of File[]

amjedidiah committed:
feat(preview): implemented universal portal-based file preview

- Replaced DOCX-specific preview with universal file preview component
- Added modal preview support for all file types
- Fixed file name truncation in preview cards

amjedidiah committed:
chore(app): cleaned up unused files

amjedidiah committed:
feat(events): implemented comprehensive upload event system

- Added file handling events (onFileClick, onFilesChange, onFileRemove)
- Added upload lifecycle events (onUpload, onCompletedUpload, onAllCompleted, onUploadFail)
- Added progress tracking events (onFileProgress, onTotalProgress)
- Added drag-and-drop events (onDragOver, onDragLeave, onDrop)
- Added integration selection event (onClick)

The event system provides granular control and visibility into:
- File selection and removal
- Upload progress per file and overall
- Upload success/failure states
- Drag-and-drop interactions
- Integration method selection
- Updates

Demonstrates all events in Storybook with status updates and console logs

amjedidiah committed:
feat(app): implemented event handlers for mini uploader

BSalaeddin commented on src/types/BaseConfigs.ts:
### It might be worth including a few additional events to cover more use cases and enhance flexibility. Here are some suggestions:

`onFileTypeMismatch`: Called when a file's type doesn’t match the allowed types, useful for providing feedback about supported formats.

`onCancelUpload`: Handles user-initiated upload cancellations, providing better control over the upload flow.

###  We could benefit from slight refinements to reduce ambiguity and ensure consistency.

`onClick`
Issue: The name is too generic and doesn’t describe its purpose (related to integration type selection).
Suggestion: Rename it to `onIntegrationClick `or `onIntegrationSelected `to clarify its intent.

`onUpload`
Issue: Could be misinterpreted as a trigger for when the entire upload process starts, but it actually refers to individual file uploads.
Suggestion: Rename it to `onFileUploadStart `for clarity.


`onCompletedUpload`

Issue: The name doesn’t specify that it’s for a single file’s upload completion.
Suggestion: Rename it to `onFileUploadComplete `to indicate it's for individual files.
`onAllCompleted`

Issue: Slightly ambiguous; could be confused with completing all operations, not just uploads.
Suggestion: Rename it to `onAllUploadsComplete `to make it clear it relates to uploads.

`onUploadFail`

Issue: The name is fine but could be improved for consistency with other events.
Suggestion: Rename it to `onFileUploadFail`.


`onTotalProgress`

Issue: Clear for aggregate progress but could use a slight clarification for upload context.
Suggestion: Rename it to `onTotalUploadProgress`.

Drag events could be prefixed with onFile (e.g., `onFileDragOver`) for better specificity.

amjedidiah replied:
Thanks @BSalaeddin , done

amjedidiah committed:
refactor(app): name refinements to reduce ambiguity and ensure consistency

per review, changed:
- `onClick` to `onIntegrationClick`
- `onUpload` to `onFileUploadStart`
- `onCompletedUpload` to `onFileUploadComplete`
- `onAllCompleted` to `onAllUploadsComplete`
- `onUploadFail` to `onFileUploadFail`
- `onTotalProgress` to `onTotalUploadProgress`
- `onDragOver` to `onFileDragOver`
- `onDragLeave` to `onFileDragLeave`
- `onDrop` to `onFileDrop`

fix failing build caused by `src/lib/uploadObject.ts`

amjedidiah committed:
feat: added onFileTypeMismatch and onCancelUpload events

- Added onFileTypeMismatch event to handle file type validation failures
- Added onCancelUpload event for upload cancellation handling
- Updated useProgress hook to properly handle upload abortion
- Exposed cancelUpload method through component ref

MedAmine1212 commented on src/UpupUploader.tsx:
```diff
@@ -60,6 +64,13 @@ export type UploadFilesRef = {
+        id: `${file.name}-${file.size}-${file.lastModified}-${Math.random()}`,
+    })
+}
+
```
Change math.random() with an actual UUID.
(please make sure to not use math.random anywhere)

amjedidiah replied:
Thanks @MedAmine1212, done

amjedidiah committed:
refactor(app): swapped `Date.now` and `Math.random` calls for more accurate values from `uuid` call

amjedidiah committed:
Merge branch 'master' into CU-86b2qw06r_feat-add-upload-event-listeners

amjedidiah committed:
fix(app): prettier fixes
